### ISSUE TITLE
Gemini API Provider

### ISSUE THREAD

**Issue by leftmove** [arshad-yaseen/monacopilot#78]
I'm currently working on a Gemini API (Google) provider for this project. I'm doing it for a personal project of mine, but I figure it would be useful if I made a pull request with my code.

Would this be something the project would benefit from?

**Comment by leftmove** [comment]
A couple things I noticed that might be useful to change (outside of the actual Google provider).

- Consider creating types for each individual provider. Right now, you can mix-and-match different LLM models and providers - even if those providers don't match to that model. I assume it will give an error later, but TypeScript should give an error natively, no? Here's what I mean in code.

```typescript

/* Change this */
export interface CopilotOptions {
  provider?: CopilotProvider;
  model?: CopilotModel | CustomCopilotModel;
}

/* To this */

export interface CopilotOptions {
  provider?: CopilotProvider;
  options?: AnthropicOptions | OpenAIOptions | GroqOptions | GoogleOptions
}


export interface AnthropicOptions { /** You could add a parent interface to extend off of to maintain consistency among  different options interfaces **/

  provider: "anthropic"
  model: AnthropicModel

}

/* ... continue for OpenAI, Groq and Gemini models */

```

- Give more info in error messages. To trace errors when developing, I had to use console log and trace down the specific HTTP library used and `console.log` my error message there. This could be solved by including the actual error message given by the different APIs.

```typescript

/* Current implementation for error handling */

const request = async <
  ResponseType,
  BodyType = undefined,
  MethodType extends Method = Method,
>(
  url: string,
  method: MethodType,
  options: RequestOptions<BodyType, MethodType> = {},
): Promise<ResponseType> => {
  const headers = {
    'Content-Type': 'application/json',
    ...options.headers,
  };

  const body =
    method === 'POST' && options.body
      ? JSON.stringify(options.body)
      : undefined;

  const response = await fetch(url, {
    method: method,
    headers,
    body,
    signal: options.signal,
  });

  if (!response.ok) {
    throw new Error(
      `${response.statusText || options.fallbackError || 'Network error'}`,
    );
  }

  return response.json() as Promise<ResponseType>;
};

/* Possibly improved implementation */

const request = async <
  ResponseType,
  BodyType = undefined,
  MethodType extends Method = Method,
>(
  url: string,
  method: MethodType,
  options: RequestOptions<BodyType, MethodType> = {},
): Promise<ResponseType> => {
  const headers = {
    'Content-Type': 'application/json',
    ...options.headers,
  };

  const body =
    method === 'POST' && options.body
      ? JSON.stringify(options.body)
      : undefined;

  const response = await fetch(url, {
    method: method,
    headers,
    body,
    signal: options.signal,
  });

  if (!response.ok) {
	const data = "\n" + JSON.stringify(await response.json() || '', null, 2)
    throw new Error(
      `${response.statusText || options.fallbackError || 'Network error'}${data}`, /* Important part is including the `response.json()`, formatting can be whatever.
    );
  }

  return response.json() as Promise<ResponseType>;
};

```

Let me know if any of these changes could be useful - I will be happy to help implement them!

Thanks for maintaining this awesome project!

**Comment by arshad-yaseen** [comment]
Thank you! The Gemini API was exactly what I was thinking about adding to Monacopilot, and you made it happen. The suggestions are great, and I can implement them in your PR. Again, thank you!

**Comment by arshad-yaseen** [comment]
@leftmove Merged your branch and released `v0.15.0` with Gemini provider 🚀 Thank you for your contribution.
